pkg/listeners/listeners_test.go

```diff
@@ -0,0 +1,142 @@
+// +build linux,cgo
+
+package listeners
+
+import (
+	"crypto/tls"
+	"flag"
+	"net"
+	"os"
+	"strconv"
+	"testing"
+
+	apiserver "github.com/docker/docker/api/server"
+	"github.com/docker/docker/cmd/dockerd/hack"
+	"github.com/docker/docker/dockerversion"
+	"github.com/docker/go-connections/tlsconfig"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var rootEnabled = false
+
+func init() {
+	flag.BoolVar(&rootEnabled, "test.root", false, "enable tests that require root")
+}
+func RequiresRoot(t *testing.T) {
+	if !rootEnabled {
+		t.Skip("skipping test that requires root")
+		return
+	}
+	require.Equal(t, 0, os.Getuid(), "This test must be run as root.")
+}
+func TestInitForUnix(t *testing.T) {
+	RequiresRoot(t)
+
+	proto := "unix"
+	addr := "/tmp/example.sock"
+	tlsConfig := tlsconfig.ServerDefault()
+
+	garbageGroupName := "garbage"
+
+	ls, serverConfig, err := commonConfig(proto, addr, garbageGroupName, tlsConfig)
+	assert.Equal(t, "group "+garbageGroupName+" not found", err.Error())
+
+	ls, serverConfig, err = commonConfig(proto, addr, "root", tlsConfig)
+	require.NoError(t, err)
+
+	serverCommonConfig(ls, proto, addr, serverConfig)
+
+	errClient := validateClient(proto, addr)
+	require.NoError(t, errClient)
+}
+
+func TestInitForTCP(t *testing.T) {
+	proto := "tcp"
+	addr := "127.0.0.1:4004"
```

Comment:
Is it possible to use `:0` to get a dynamic port allocation? Otherwise the test may be flaky if something else on the system happens to be using this port.